plotter/conrec.go

```diff
@@ -0,0 +1,177 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package plotter
+
+import (
+	"math"
+)
+
+type point struct {
+	X, Y float64
+}
+
+type line struct {
+	p1, p2 point
+}
+
+func sect(h, p [5]float64, v1, v2 int) float64 {
+	return (h[v2]*p[v1] - h[v1]*p[v2]) / (h[v2] - h[v1])
+}
+
+// conrecLine performs an operation with a line at a given height derived
+// from data over the 2D box interval (i, j) to (i+1, j+1).
+type conrecLine func(i, j int, l line, height float64)
+
+// conrec is a Go translation of the C version of CONREC by Paul Bourke:
+// http://paulbourke.net/papers/conrec/conrec.c
+//
+// conrec takes g, an m×n grid function, a sorted slice of contour heights
+// and a conrecLine function.
+//
+// For full details of the algorithm, see the paper at
+// http://paulbourke.net/papers/conrec/
+func conrec(g GridXYZ, heights []float64, fn conrecLine) {
+	var (
+		p1, p2 point
+
+		h      [5]float64
+		sh     [5]int
+		xh, yh [5]float64
+
+		im = [4]int{0, 1, 1, 0}
+		jm = [4]int{0, 0, 1, 1}
+
+		cases = [3][3][3]int{
+			{{0, 0, 8}, {0, 2, 5}, {7, 6, 9}},
+			{{0, 3, 4}, {1, 0, 1}, {4, 3, 0}},
```

Comment:
The middle point of this is "3" in the conrec.c algorithm, but is 0 here. Is that intentional?